File: 프로그래머스/2/131704. 택배상자/택배상자.js
function solution(order) {
    var answer = 0;
    
    const len = order.length;
    let main = Array.from({length : len}, (v,i) => i+1);
    main.sort((a,b) => b-a);
    let sub = [];
    let idx =main.length -1;
    
    for(const item of order) {
        
        
        if(main.length === 0 || main[idx] > item) {
            if(sub[sub.length -1] === item) {
                sub.pop();
                answer +=1;
                continue;
            } else break;
        }
        
        while(main[idx] < item) {
            sub.push(main[idx]);
            main.pop();
            idx -=1;
        }
        
        if(main[idx] === item) {
            answer +=1;
            idx -=1;
            main.pop();
        } else break;
    }
    
    return answer;
}

File: 프로그래머스/2/131704. 택배상자/택배상자.cpp
#include <string>
#include <vector>
#include <stack>
#include <queue>

using namespace std;

int solution(vector<int> order) {
    int answer = 0;
    
    int len = order.size();
    
    queue<int> q;
    
    for(int i=1; i<=len ; i++)
        q.push(i);
    
    stack<int> s;
    
    for(auto item : order)
    {
        if(q.empty() || q.front() > item)
        {
            if(s.top() != item) break;
            else
            {
                answer +=1;
                s.pop();
                continue;
            }
        }
        
        while(q.front() < item)
        {
            s.push(q.front());
            q.pop();
        }
        
        if(q.front() == item)
        {
            answer +=1;
            q.pop();
        } else break;
    }
    
    return answer;
}